pkg/v2/tkr/webhook/cluster/tkr-resolver/cluster/cluster.go

```diff
@@ -6,16 +6,335 @@ package cluster
+	"k8s.io/apimachinery/pkg/selection"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 
+	runv1 "github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha3"
 	"github.com/vmware-tanzu/tanzu-framework/pkg/v2/tkr/resolver"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v2/tkr/resolver/data"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v2/tkr/util/osimage"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v2/tkr/util/version"
 )
 
 type Webhook struct {
-	TKRResolver resolver.Resolver
+	TKRResolver resolver.CachingResolver
+	Log         logr.Logger
+	Client      client.Client
+	decoder     *admission.Decoder
 }
 
-func (cw *Webhook) Handle(_ context.Context, req admission.Request) admission.Response { // nolint:gocritic // suppress linter error: hugeParam: req is heavy (400 bytes); consider passing by pointer (gocritic)
-	return admission.Allowed("Everything is fine :) req.Name: '" + req.Name + "'")
+func (cw *Webhook) InjectDecoder(decoder *admission.Decoder) error {
+	cw.decoder = decoder
+	return nil
+}
+
+func (cw *Webhook) Handle(ctx context.Context, req admission.Request) admission.Response { // nolint:gocritic // suppress linter error: hugeParam: req is heavy (400 bytes); consider passing by pointer (gocritic)
+	cluster := &clusterv1.Cluster{}
+	if err := cw.decoder.Decode(req, cluster); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	cw.Log.Info("resolving cluster", "namespace", cluster.Namespace, "name", cluster.Name)
+	clusterClass, err := cw.getClusterClass(ctx, cluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) { // cluster is classy, but its ClusterClass is nowhere to be found
+			return admission.Denied("ClusterClass not found")
+		}
+		cw.Log.Error(err, "error getting ClusterClass")
+		return admission.Errored(http.StatusBadGateway, errors.Wrap(err, "error getting ClusterClass"))
+	}
+
+	if err := cw.ResolveAndSetMetadata(cluster, clusterClass); err != nil {
+		return admission.Denied(err.Error())
+	}
+
+	return success(&req, cluster)
+}
+
+// ResolveAndSetMetadata uses cw.TKRResolver and injects resolved metadata into the provided cluster.
+// Pre-reqs: cluster != nil && clusterClass != nil
+func (cw *Webhook) ResolveAndSetMetadata(cluster *clusterv1.Cluster, clusterClass *clusterv1.ClusterClass) error {
+	query, err := cw.constructQuery(cluster, clusterClass)
+	if err != nil {
+		return err
+	}
+	if isEmpty(query) { // skipping resolution entirely
+		return nil
+	}
+
+	result := cw.TKRResolver.Resolve(query)
+
+	isUnresolvedCP := isUnresolved(result.ControlPlane)
+	unresolvedMDs := unresolvedMachineDeployments(result)
+	if isUnresolvedCP || len(unresolvedMDs) != 0 {
+		return &errUnresolved{
+			query:   query,
+			result:  result,
+			cluster: cluster,
+			cp:      isUnresolvedCP,
+			mds:     unresolvedMDs,
+		}
+	}
+
+	setMetadata(result, cluster)
+	cw.adjustKubernetesVersion(cluster)
```

Comment:
Nice! (It would enable the users to give the kubernetes version prefix and it would set the resolved kubernetes version with accepted format)